src/kakarot/stack.cairo

```
@@ -77,30 +91,110 @@ namespace Stack {
+        // Read and Copy the elements on an array
+
+        // Generate an array of Uint256* to return
+        let (stack_word_dict) = dict_copy(
+            stack_word_dict=stack_word_dict, stack_len=position_zero, n=n * 2, output=new_elements
+        );
+
+        // Return Stack with updated Len
+        let popped_len = 2 * n;
+        return (
+            new model.Stack(
+            stack_word_dict_start=self.stack_word_dict_start,
+            stack_word_dict=stack_word_dict,
+            stack_16bytes_len=self.stack_16bytes_len - popped_len,
+            ),
+            new_elements,
+        );
+    }
+
+    func dict_copy{range_check_ptr}(
+        stack_word_dict: DictAccess*, stack_len: felt, n: felt, output: Uint256*
+    ) -> (stack_word_dict: DictAccess*) {
+        if (n == 0) {
+            return (stack_word_dict=stack_word_dict);
+        }
+
+        // Get Low and High of element at position N
+        let (el_high) = dict_read{dict_ptr=stack_word_dict}(stack_len - n);
+        let (el_low) = dict_read{dict_ptr=stack_word_dict}(stack_len - n + 1);
+
+        // Save Uint256 value in array
+        let n_index = n / 2 - 1;
+        assert output[n_index] = Uint256(low=el_low, high=el_high);
+
+        return dict_copy(
+            stack_word_dict=stack_word_dict, stack_len=stack_len, n=n - 2, output=output
+        );
+    }
+
+    // @notice Pop an element from the stack.
+    // @param self - The pointer to the stack.
+    // @return The new pointer to the stack.
+    // @return The popped element.
+    func pop{range_check_ptr}(self: model.Stack*) -> (new_stack: model.Stack*, element: Uint256) {
+        let stack_word_dict = self.stack_word_dict;
+        let position_zero = self.stack_16bytes_len;
+        // Check if stack will underflow
+
+        if (position_zero == 0) {
+            with_attr error_message("Kakarot: StackUnderflow") {
+                assert 1 = 0;
+            }
+        }
```

same as before, why not asserting what we actually want to check?

---

The assert already happens. if it's 0, we want to raise an error -> revert.
Though, we can use assert_non_zero() here instead.

---

Although, inlining it is better, especially for an op that repeats a lot like stack stuff

---

not sure to understand what you mean by inlining, I was talking about your proposed solution of doing:

```
with_attr error_message("Kakarot: StackUnderflow") {
      assert_non_zero(position_zero);
}
```

---

Yeah, it's a bit less efficient, since you have a call and ret instructions for the good flow, instead of 1 instruction in the if